url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/example/GraphEdX.java

```diff
@@ -0,0 +1,642 @@
+					viewPort.setView(graph);
+				} else {
+					// Best effort...
+					parent.add(graph);
+				}
+				// And reinstalls the listener
+				installListeners(graph);
+			}
+		}
+	}
+
+	public void deserializeGraph() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showOpenDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marqueeHandler = graph.getMarqueeHandler();
+			try {
+				uninstallListeners(graph);
+				parent.remove(graph);
+				ObjectInputStream in = new ObjectInputStream(
+						new BufferedInputStream(new FileInputStream(fileChooser
+								.getSelectedFile())));
+				graph = (JGraph) in.readObject();
+				// Take the marquee handler from the original graph and
+				// use it in the new graph as well.
+				graph.setMarqueeHandler(marqueeHandler);
+				// Adds the component back into the component hierarchy
+				if (parent instanceof JViewport) {
+					JViewport viewPort = (JViewport) parent;
+					viewPort.setView(graph);
+				} else {
+					// Best effort...
+					parent.add(graph);
+				}
+				// graph.setMarqueeHandler(previousHandler);
+				// And reinstalls the listener
+				installListeners(graph);
+			} catch (Exception e) {
+				e.printStackTrace();
+				JOptionPane.showMessageDialog(graph, e.getMessage(), "Error",
+						JOptionPane.ERROR_MESSAGE);
+			}
+		}
+	}
+
+	public void saveFile() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showSaveDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			XMLEncoder encoder;
+			Container parent = graph.getParent();
+			try {
+				uninstallListeners(graph);
+				parent.remove(graph);
+				encoder = new XMLEncoder(new BufferedOutputStream(
+						new FileOutputStream(fileChooser.getSelectedFile())));
+				configureEncoder(encoder);
+				encoder.writeObject(graph);
+				encoder.close();
+			} catch (Exception e) {
+				JOptionPane.showMessageDialog(graph, e.getMessage(), "Error",
+						JOptionPane.ERROR_MESSAGE);
+			} finally {
+				// Adds the component back into the component hierarchy
+				if (parent instanceof JViewport) {
+					JViewport viewPort = (JViewport) parent;
+					viewPort.setView(graph);
+				} else {
+					// Best effort...
+					parent.add(graph);
+				}
+				// And reinstalls the listener
+				installListeners(graph);
+			}
+		}
+	}
+
+	public void openFile() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showOpenDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marqueeHandler = graph.getMarqueeHandler();
+			try {
+				uninstallListeners(graph);
+				parent.remove(graph);
+				XMLDecoder decoder = new XMLDecoder(new BufferedInputStream(
+						new FileInputStream(fileChooser.getSelectedFile())));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *BufferedInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 440 evaluates to true. Ensure the object returned by *BufferedInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).